presto-native-execution/presto_cpp/metrics/PrometheusStatsReporter.cpp

```diff
@@ -0,0 +1,170 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+ */
+
+#include "PrometheusStatsReporter.h"
+#include "velox/common/base/StatsReporter.h"
+#include "prometheus/client_metric.h"
+#include "prometheus/counter.h"
+#include "prometheus/summary.h"
+#include "prometheus/gauge.h"
+
+
+namespace facebook::presto {
+
+void PrometheusStatsReporter::registerMetricExportType(folly::StringPiece key, facebook::velox::StatType statType) const {
+  registerMetricExportType(key.start(), statType);
+}
+
+
+void PrometheusStatsReporter::registerMetricExportType(const char* key, velox::StatType statType) const {
+  std::string keyStr = sanitizeMetricName(key);
+  statTypeMap.insert({keyStr, statType});
+  LOG(INFO) << "Registered metric: " << keyStr << " with type: " << statTypeToString(statType);
+
+  switch (statType) {
+    case velox::StatType::AVG:{
+      registerSummaryMetricForAverageStatType(keyStr);
+      break;
+    }
+    case velox::StatType::SUM: {
+      registerGaugeMetric(keyStr);
+      break;
+    }
+    case velox::StatType::COUNT: {
+      registerCounterMetric(keyStr);
+      break;
+    }
+    case velox::StatType::RATE:
+      // ToDo: RATE Type is not yet used anywhere
+    default:
+      LOG(WARNING) << "Unsupported stat type for key: " << keyStr;
+      break;
+  }
+}
+
+
+
+void PrometheusStatsReporter::addMetricValue(const char* key, size_t value)
+    const {
+  addMetricValue(std::string(key), value);
+}
+
+void PrometheusStatsReporter::addMetricValue(folly::StringPiece key, size_t value)
+    const {
+  addMetricValue(key.str(), value);
+}
+
+void PrometheusStatsReporter::addMetricValue(const std::string& key, size_t value) const {
+  std::string keyStr = sanitizeMetricName(key);
+  auto statTypeIt = statTypeMap.find(keyStr);
+  if (statTypeIt == statTypeMap.end()) {
+    LOG(WARNING) << "Metric not registered: " << keyStr;
+    return;
+  }
+
+  switch (statTypeIt->second) {
+    case velox::StatType::AVG: {
+      updateSummaryMetric(keyStr, value);
+      break;
+    }
+    case velox::StatType::SUM: {
+      updateGaugeMetric(keyStr, value);
+      break;
+    }
+    case velox::StatType::COUNT: {
+      updateCounterMetric(keyStr, value);
+      break;
+    }
+    case velox::StatType::RATE:
+      // ToDo: RATE Type is not yet used anywhere
+    default:
+      LOG(WARNING) << "Unsupported stat type for key: " << keyStr;
+      break;
+  }
+}
+
+void PrometheusStatsReporter::registerSummaryMetricForAverageStatType(const std::string& keyStr) const {
+  auto& summaryFamily = prometheus::BuildSummary()
+                            .Name(keyStr)
+                            .Help("Summary for " + keyStr)
+                            .Register(*registry);
+  auto& summary = summaryFamily.Add({{"cluster", cluster_}, {"node", node_}, {"host", host_}},
+                                    prometheus::Summary::Quantiles{{0.5, 0.05}});
```

Comment:
0.5 quantile is not the average, instead it is the median value. This is as good as reporting instantaneous value.

Reply:
![image](https://github.com/prestodb/presto/assets/19339828/49fc8034-ec20-4e15-bb69-9a9b9b6293ba)
It has sum and count so avg can be computed as well